lib/ukboot/early_init.c

```diff
@@ -0,0 +1,150 @@
+/* Copyright (c) 2024, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+#include <uk/argparse.h>
+#include <uk/assert.h>
+#include <uk/boot/earlytab.h>
+#include <uk/essentials.h>
+#include <uk/plat/common/bootinfo.h>
+#include <uk/prio.h>
+
+#if CONFIG_LIBUKLIBPARAM
+#include <uk/libparam.h>
+#endif /* CONFIG_LIBUKLIBPARAM */
+
+char *boot_argv[CONFIG_LIBUKBOOT_MAXNBARGS];
+int boot_argc;
+
+/* Copy of bi->cmdline, as the boot protocol buffer may
+ * be unmapped by paged memory init.
+ */
+static char cmdline[CONFIG_LIBUKBOOT_CMDLINE_MAX_LEN];
+static __sz cmdline_len;
+
+/* Mutable version of the cmdline. This is tokenized by uk_argnparse().
+ * Keep this as a global as `argv` points to tokens in this string.
+ */
+static char cmdline_argv[CONFIG_LIBUKBOOT_CMDLINE_MAX_LEN];
+
+#define uk_boot_earlytab_foreach(itr)			\
+	for ((itr) = (struct uk_boot_earlytab_entry *)	\
+			uk_boot_earlytab_start;		\
+	     (itr) < &(uk_boot_earlytab_end);		\
+	     (itr)++)
+
+#if CONFIG_LIBUKLIBPARAM
+static int init_libparam(int *argc, char *argv[])
+{
+	int rc;
+
+	UK_ASSERT(argc);
+	UK_ASSERT(argv);
+
+	/*
+	 * First, we scan if we can find the stop sequence in the kernel
+	 * cmdline. If not, we assume that there are no uklibparam arguments in
+	 * the command line.
+	 * NOTE: argv[0] contains the kernel/program name that we need to
+	 *       hide from the parser.
+	 */
+	rc = uk_libparam_parse(*argc - 1, &argv[1], UK_LIBPARAM_F_SCAN);
+	if (rc > 0 && rc < (*argc - 1)) {
+		/* In this case, we did successfully scan for uklibparam
+		 * arguments and stop sequence is at rc < (argc - 1).
+		 */
+		/* Run a second pass for parsing */
+		rc = uk_libparam_parse(*argc - 1, &argv[1],
+				       UK_LIBPARAM_F_USAGE);
+		if (unlikely(rc < 0)) /* go down on errors (including USAGE) */
+			return -EINVAL;
+
+		/* Drop uklibparam parameters from argv but keep argv[0].
+		 * We are going to replace the stop sequence with argv[0].
+		 */
+		rc += 1; /* include argv[0]; we use rc as idx to stop seq */
+		*argc -= rc;
+		argv[rc] = argv[0];
+		argv = &argv[rc];
+	}
+	return 0;
+}
+#endif /* CONFIG_LIBUKLIBPARAM */
+
+static int uk_boot_early_init_parse_cmdl(struct ukplat_bootinfo *bi)
+{
+	int rc __maybe_unused;
+	const char *cmdl;
+
+	UK_ASSERT(bi);
+
+	if (!bi->cmdline_len) {
+		cmdl = CONFIG_UK_NAME;
+		cmdline_len = sizeof(CONFIG_UK_NAME);
+	} else {
+		cmdl = (const char *)bi->cmdline;
+		cmdline_len = bi->cmdline_len;
+	}
+
+	if (unlikely(cmdline_len > ARRAY_SIZE(cmdline))) {
+		rc = -E2BIG;
+		uk_pr_err("Could not parse cmdline (%d)\n", rc);
+		return rc;
+	}
+
+	/* Update bi->cmdline */
+	strncpy(cmdline, cmdl, ARRAY_SIZE(cmdline));
+	cmdline[cmdline_len] = 0;
+	bi->cmdline = (__u64)cmdline;
+
+	/* Break down cmdline to argv */
+	strncpy(cmdline_argv, cmdl, ARRAY_SIZE(cmdline));
+	cmdline_argv[cmdline_len] = 0;
```

Comment:
I'd prefer to stay with the target data type: Can we place `'\0'` here, instead of `0`?